flocker/testtools/__init__.py

```diff
@@ -868,6 +870,76 @@ def __str__(self):
         return base + " and output:\n" + lines
 
 
+class _LoggingProcessProtocol(ProcessProtocol):
+    """
+    A ``ProcessProtocol`` that stores output from the subprocess and also logs
+    output as its received as eliot messages.
+
+    Intended to be used by ``logged_run_process``.
+    """
+
+    def __init__(self, deferred):
+        """
+        Construct a ``_LoggingProcessProtocol``.
+
+        :param deferred: A ``Deferred`` that will fire with ``(reason, output)``
+            when the process ends, where ``reason`` is a ``Failure`` with the
+            reason for the process ending (see ``IProcessProtocol``), and
+            ``output`` are the bytes outputted by the process (both to stdout
+            and stderr).
+        """
+        self._deferred = deferred
+        self._output_buffer = StringIO()
+
+    def _log_output(self, data):
+        """
+        Log ``data`` as an eliot message.
+        """
+        # TODO: This should probably be within some sort of action context.
+        # TODO: Do we want to line buffer this?
+        Message.new(output=data).write()
+
+    def outReceived(self, data):
+        self._output_buffer.write(data)
+        self._log_output(data)
+
+    def errReceived(self, data):
+        self._output_buffer.write(data)
+        self._log_output(data)
+
+    def processExited(self, reason):
+        self._deferred.callback((reason, self._output_buffer.getvalue()))
+
+
+def logged_run_process(reactor, command):
+    """
+    Run a child process, and log the output as we get it.
+
+    :param reactor: An ``IReactorProcess`` to spawn the process on.
+    :param command: An argument list specifying the child process to run.
+
+    :return: A ``Deferred`` that calls back with ``_ProcessResult`` if the
+        process exited successfully, or errbacks with
+        ``_CalledProcessError`` otherwise.
+    """
+    d = Deferred()
+    protocol = _LoggingProcessProtocol(d)
+    reactor.spawnProcess(protocol, command[0], command)
+
+    def process_ended((reason, output)):
+        status = reason.value.status
+        if status:
+            raise _CalledProcessError(
+                returncode=status, cmd=command, output=output)
+        return _ProcessResult(
+            command=command,
+            status=reason.value.status,
```

Comment:
You can use `status=status` here (technically, you could even use `status=0` but that's less obvious).